src/matrix-to.js

```diff
@@ -25,17 +25,211 @@ export const baseUrl = `https://${host}`;
+// servers fail them (in terms of joining) are somewhat expected to hunt
+// down the person who gave them the link to ask for a participating server.
+// The receiving user can then manually append the known-good server to
+// the list and magically have the link work.
+
+export class RoomPermalinkCreator {
+    constructor(room) {
+        this._room = room;
+        this._highestPlUserId = null;
+        this._populationMap = null;
+        this._bannedHostsRegexps = null;
+        this._allowedHostsRegexps = null;
+        this._serverCandidates = null;
+
+        this.onMembership = this.onMembership.bind(this);
+        this.onRoomState = this.onRoomState.bind(this);
+    }
+
+    load() {
+        this._updateAllowedServers();
+        this._updateHighestPlUser();
+        this._updatePopulationMap();
+        this._updateServerCandidates();
+    }
+
+    start() {
+        this.load();
+        this._room.on("RoomMember.membership", this.onMembership);
+        this._room.on("RoomState.events", this.onRoomState);
+    }
+
+    stop() {
+        this._room.removeListener("RoomMember.membership", this.onMembership);
+        this._room.removeListener("RoomState.events", this.onRoomState);
+    }
+
+    forEvent(eventId) {
+        const roomId = this._room.roomId;
+        const permalinkBase = `${baseUrl}/#/${roomId}/${eventId}`;
+        return `${permalinkBase}${encodeServerCandidates(this._serverCandidates)}`;
+    }
+
+    forRoom() {
+        const roomId = this._room.roomId;
+        const permalinkBase = `${baseUrl}/#/${roomId}`;
+        return `${permalinkBase}${encodeServerCandidates(this._serverCandidates)}`;
+    }
+
+    onRoomState(event) {
+        switch (event.getType()) {
+            case "m.room.server_acl":
+                this._updateAllowedServers();
+                this._updateHighestPlUser();
+                this._updatePopulationMap();
+                this._updateServerCandidates();
+                return;
+            case "m.room.power_levels":
+                this._updateHighestPlUser();
+                this._updateServerCandidates();
+                return;
+        }
+    }
+
+    onMembership(evt, member, oldMembership) {
+        const userId = member.userId;
+        const membership = member.membership;
+        const serverName = getServerName(userId);
+        const hasJoined = oldMembership !== "join" && membership === "join";
+        const hasLeft = oldMembership === "join" && membership !== "join";
+
+        if (hasLeft) {
+            this._populationMap[serverName]--;
+        } else if (hasJoined) {
+            this._populationMap[serverName]++;
+        }
+
+        this._updateHighestPlUser();
+        this._updateServerCandidates();
+    }
+
+    _updateHighestPlUser() {
+        const plEvent = this._room.currentState.getStateEvents("m.room.power_levels", "");
+        const content = plEvent && plEvent.getContent();
+        if (content) {
+            const users = content.users;
```

Comment:
This will fail because `content` is now a boolean

The block of code could be:
```
const plEvent = ...;
if (!plEvent) {
  this.highestPlUser = null;
  return;
}
const content = plEvent.getContent(); // btw, this will always return something so the null check might not be needed
if (content) {
...
```

Reply:
It shouldn't be, as `&&` doesn't do any coercion:
```
const a = {b:4};
let b = a && a.b // b is 4
let c = a && a.c // b is undefined
```
But I'll change it if the code is confusing like this.

Reply:
javascript is the worst >:(

Changing it, or a comment, would be great